Add neon_broker binary.

arssher opened the pull request:
Which ought to replace neon_broker. This patch only adds the binary and adjusts Dockerfile to include it; subsequent ones will add deploy of helm chart and the actual replacement.

It is a simple and fast pub-sub message bus. In this patch only safekeeper message is supported, but others can be easily added.

Compilation now requires protoc to be installed. Installing protobuf-compiler package is fine for Debian/Ubuntu.

ref
https://github.com/neondatabase/neon/pull/2733
https://github.com/neondatabase/neon/issues/2394

koivunej commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+        let mut cleanup_subs_to_all = false;
+        for sub in self.subs_to_all.iter() {
+            match sub.try_send(msg.clone()) {
+                Err(TrySendError::Full(_)) => {
+                    trace!("dropping message, channel is full");
+                }
```
This Drop when full behaviour makes me wonder if per publisher [tokio::sync::broadcast](https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html) channel would had been more applicable. On being full it "drops" oldest messages and resumes from oldest still reachable, compared to this which drops the later messages but continues to send older, skipping some in the middle.

Memory-wise it would allow all subscribers to follow a single queue instead of having their own, removing some cleanup lines at the cost of having more complexity inside subscriber's `try_stream!`, unsure how to handle the `subs_to_all` at least right away... EDIT: I don't think that's an issue since one subscriber only subscribes to all or specific timelines.

Reading this more I think it would help with not needing to be concerned about removal or even book keeping of subscribers.

If you already considered it but it cannot work for some reason I cannot see, could a comment be added somewhere?

petuhovskiy replied:
Nice, [tokio::sync::broadcast](https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html) looks like a ready implementation of what a broker does, i.e. Pub/Sub. When we will rewrite the broker code, we should definitely consider using tokio::sync::broadcast.

arssher replied:
Cool, I didn't know it has support for dropping slow receivers messages, will try to switch.

koivunej commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+        // transform rx into stream with item = Result, as method result demands
+        let output = async_stream::try_stream! {
+            while let Some(info) = subscriber.sub_rx.recv().await {
+                    yield info
+                }
```
Small nit about the formatting since rustfmt doesn't touch these blocks.

```suggestion
                yield info
            }
```

koivunej commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+// provided channel, or forget about it, releasing memory.
+#[derive(Clone)]
+enum SubAnnounce {
+    AddAll(Sender<SafekeeperTimelineInfo>), // add subscription to all timelines
+    AddTimeline(TenantTimelineId, SubSender), // add subsciption to the specific timeline
+    RemoveTimeline(TenantTimelineId, SubId), // remove subscription to the specific timeline
+                                            // RemoveAll is not needed as publisher will notice closed channel while
+                                            // trying to send the next message.
```
Since this is new code a nit: would prefer to see rustdoc friendly comments before the variants.

```suggestion
    /// Add subscription to all timelines
    AddAll(Sender<SafekeeperTimelineInfo>),
    /// Add subscription to the specific timeline
    AddTimeline(TenantTimelineId, SubSender),
    /// Remove subscription to the specific timeline
    RemoveTimeline(TenantTimelineId, SubId),
    
    // RemoveAll is not needed as publisher will notice closed channel while
    // trying to send the next message.
```

koivunej reviewed (commented):
I think this is looking good, with very nice comments. Added some comment style nits since this is new code, and a question about tokio::sync::broadcast usage instead.

Could some sort of smoke test be added, for example that a published update is seen both by an "all" subscriber and a specific tenant-timeline subscriber?

SomeoneToIgnore commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+        for sub in self.subs_to_all.iter() {
+            match sub.try_send(msg.clone()) {
+                Err(TrySendError::Full(_)) => {
+                    trace!("dropping message, channel is full");
```
I think we did not change that and the other `trace!` to warn (or added the metrics, or something else)
See https://github.com/neondatabase/neon/pull/2733#discussion_r1012156020

SomeoneToIgnore commented on broker/readme.md:
````diff
@@ -0,0 +1,4 @@
+```
````
The readme is still odd, see https://github.com/neondatabase/neon/pull/2733#discussion_r1011764691 
I think, current commit message is better than this readme 🙂 

> Compilation now requires protoc to be installed

This belongs to the main readme, where we list all things needed for compilation: https://github.com/neondatabase/neon/blame/c4b417ecdbc030f245673362dc23fbf4f279180e/README.md#L37

I think we should change all installation instructions, for Mac, the package is named `protobuf-c`

SomeoneToIgnore reviewed (commented):
Not sure when the "formal" addition of the tool is supposed to happen, so have transferred important unresolved comments from https://github.com/neondatabase/neon/pull/2733

IMO, we should adjust READMEs now, but ignore me if you plan to do that later.

petuhovskiy commented on broker/proto/broker.proto:
```diff
@@ -0,0 +1,44 @@
+
+message SubscribeSafekeeperInfoRequest {
+    oneof subscription_key {
+        google.protobuf.Empty all = 1; // subscribe to everything
```
I'm not sure if `Empty` inside oneof will work. As I understand, oneof requires at least one non-zero value to be passed in one of the fields and `Empty` type is always zero-value, so you probably can't set `all` field and subscribe to everything? I may be wrong here, because I've never used oneof before.

petuhovskiy replied:
It works, need to reread protobuf docs to understand what's going on.

petuhovskiy commented on broker/src/metrics.rs:
```diff
@@ -0,0 +1,22 @@
+use once_cell::sync::Lazy;
+
+pub static NUM_PUBS: Lazy<IntGauge> = Lazy::new(|| {
+    register_int_gauge!("broker_num_pubs", "Number of publications")
```
```suggestion
    register_int_gauge!("broker_active_publishers", "Number of publications")
```

petuhovskiy commented on broker/src/metrics.rs:
```diff
@@ -0,0 +1,22 @@
+
+pub static NUM_SUBS_TIMELINE: Lazy<IntGauge> = Lazy::new(|| {
+    register_int_gauge!(
+        "broker_num_subs_timeline",
```
```suggestion
        "broker_per_timeline_active_subscribers",
```

petuhovskiy commented on broker/src/metrics.rs:
```diff
@@ -0,0 +1,22 @@
+});
+
+pub static NUM_SUBS_ALL: Lazy<IntGauge> = Lazy::new(|| {
+    register_int_gauge!("broker_num_subs_all", "Number of subsciptions to all keys")
```
```suggestion
    register_int_gauge!("broker_all_keys_active_subscribers", "Number of subsciptions to all keys")
```

petuhovskiy commented:
We should add `protoc` installation instructions to the README.md in the root of the repo

petuhovskiy commented:
https://github.com/fullstorydev/grpcurl can be used as a client for the broker. For example, this query will subscribe for everything:

```
grpcurl -proto broker.proto -d '{"all":{}}' -plaintext localhost:50051 neon_broker.NeonBroker/SubscribeSafekeeperInfo
```

petuhovskiy reviewed (approved):
`neon_broker.rs` is hard to read even on the 2nd try :)
I found nothing wrong in the code, so approve from me.

LizardWizzard commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+    info!("version: {GIT_VERSION}");
+
+    let registry = Registry {
+        shared_state: Arc::new(Mutex::new(SharedState {
```
Can be #[derive(Default)] I guess

LizardWizzard commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+            select! {
+                msg = stream.next() => {
+                    match msg {
+                        Some(Ok(msg)) => {publisher.send_msg(&msg)?;},
+                        Some(Err(e)) => {return Err(e);}, // grpc error from the stream
+                        None => {break;} // closed stream
```
```suggestion
                        Some(Ok(msg)) => publisher.send_msg(&msg)?,
                        Some(Err(e)) => return Err(e), // grpc error from the stream
                        None => break // closed stream
```

LizardWizzard commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+}
+
+#[tonic::async_trait]
+impl NeonBroker for NeonBrokerImpl {
```
NIT: Maybe call it the service `BrokerService`, and then struct becomes just `Broker`, so ends up `impl BrokerService for Broker`

arssher replied:
It would make imports more verbose (already fat), so let's leave as is.

LizardWizzard replied:
> It would make imports more verbose (already fat),

But why? As I can see there wont be any difference. And no ugly `Impl` suffix as a plus.

arssher replied:
tonic generated code includes two modules, currently called `neon_broker_server` and `neon_broker_client`, these would be longer. But actually I'm fine with either variant.

LizardWizzard replied:
Longer than what? There is a change from NeonBroker to BrokerService and struct can be called just Broker after that. So overall it looks shorter to me.

`neon_broker_proto` module name  is written in our code. So can be just proto. It is in lib.rs so can be either `crate::proto` or `broker::proto`. This is part of lengthy imports discussion above

arssher replied:
> neon_broker_proto module name is written in our code. So can be just proto. 

+1

> There is a change from NeonBroker to BrokerService and struct can be called just Broker after that. So overall it looks shorter to me.

Ok, in this case it won't be longer, but I dislike `BrokerService` names as it generates imports like
```
use neon_broker::proto::broker_service_server::{BrokerService, BrokerServiceServer};
```

So perhaps just `Broker`?

arssher replied:
pondering some more, made it `BrokerService`.

LizardWizzard replied:
There have to be two entities. One is a trait that is generated by tonic and the second one is the implementation struct. So cannot be Broker in both places (or we need to alias something in `use` section). So yeah, `BrokerService` trait and `Broker` struct is IMO is a good way to describe this

LizardWizzard reviewed (dismissed):
Comments are mostly nitpicking/requests for cleanup.

I think before merge this needs to be covered with proper unittests and probably some integrational ones. Should be easy to write them in Rust.

As an idea I think it is worth considering Mutex from parking_lot instead of std one. It doesnt panic and should be faster.

It is sad to see requirement for Vec<u8> for `bytes` protobuf type. Not sure how bad it is, which alternatives should we consider at this point?

arssher commented on broker/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+name = "neon_broker_bench"
+path = "src/bin/bench.rs"
+# build benchmarking binary only if explicitly requested with '--features bench'
+required-features = ["bench"]
```
Checked out, I don't think it fits well the micro benchmarking format, for example it would be nice to print rps, not execution time. Also, #[bench] is unstable (yes, there is Criterion...). But I will add readme to the binary.

LizardWizzard replied:
Please revisit what `harness = false` does. It supports exactly this case. You write usual binary with main function that will be invoked with `cargo bench`. See this example from glommio https://github.com/DataDog/glommio/blob/master/glommio/Cargo.toml#L74

arssher replied:
Checked again, it seems impossible to run `cargo bench` without making it pass `--bench` to the binary, and clap doesn't support arguments with no value %)

arssher replied:
I was surely wrong, clap handles that, moved to [[bench]].

arssher commented on broker/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+tonic = "0.8"
+tokio = { version = "1.0", features = ["macros", "rt-multi-thread"] }
+# for exploring with tokio-console
+# tokio = { version = "1", features = ["full", "tracing"] }
```
let's just remove altogether, not sure how to require dep features depending on input features %)

arssher replied:
Tried that, everything was more or less ok, but hakari doesn't seem to follow optional features, i.e. it means tokio would be compiled with all (including tracing) features always, which is bad, so reverting.

arssher commented on broker/src/bin/bench.rs:
```diff
@@ -0,0 +1,175 @@
+    // let key = SubscriptionKey::All(());
+    let key = SubscriptionKey::TenantTimelineId(ProtoTenantTimelineId {
+        tenant_id: vec![0xFF; 16],
+        timeline_id: tli_from_u64(i),
```
If you would like to see fixed size array here, then no, protobuf doesn't support them at all. Yes, somewhat disappointing, but here I don't think this matters.

LizardWizzard replied:
Even without fixed size arrays vec shouldnt be necessary. It should be possible to populate the value from &[u8] or an iterator. There is even an issue in the library for that (not mentioning it to avoid direct linking to it). Two allocations for each message look bad for me

LizardWizzard replied:
This still looks like a problem to me. Lets ask someone else to get third opinion

arssher commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+//! subscribers.
+//!
+//! Only safekeeper message is supported, but it is not hard to add something
+//! else with templating.
```
I mean pub-sub code is agnostic to the carried message (`SafekeeperTimelineInfo` here), it is easy to extract it to template arg if we have another (e.g. similar thing for pageservers). Or alternatively we can use protobuf Any, as Arthur suggested.

LizardWizzard replied:
Do you mean generics by templating?

arssher replied:
yes

LizardWizzard replied:
Lets use proper terminology then. Its not c++ :). 

As a third option we can also just store message as bytes and let the client decide what needs to be passed here. It can be another protobuf blob, can be json.

arssher replied:
changed to generics

arssher commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+}
+
+// Args is put into OnceCell to make conf available everywhere.
+static ARGS: OnceCell<Args> = OnceCell::new();
```
Well, simplifies access, but here either is fine to me.

arssher commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,563 @@
+impl Registry {
+    // Register new publisher in shared state.
+    pub fn register_publisher(&self) -> Publisher {
+        let (announce_tx, announce_rx) = mpsc::channel(128);
```
yes, they are conceptually different; through none should be large

arssher commented on broker/src/bin/bench.rs:
```diff
@@ -0,0 +1,175 @@
+use std::time::{Duration, Instant};
+
+use clap::Parser;
+use neon_broker::neon_broker_proto::neon_broker_client::NeonBrokerClient;
```
`crate` won't work here, as this is binary crate, but I reduced that.

LizardWizzard replied:
Still do not get it why we cant call it just broker. Imports still look like that `use neon_broker::neon_broker_proto::`. Why not just broker::proto?

arssher replied:
It would be possible if lib exported these names directly in global namespace. However, there is
```
// Code generated by protobuf.
pub mod neon_broker_proto {
    include!("../proto/neon_broker.rs");
}
```

I find it useful that all generated code is put into separate module, helps differentiate with manually written one. I also saw the same structure in all tonic apps/examples I encountered.

LizardWizzard replied:
I'm ok with separate module, was just wondering why we cant call the module just `proto`. I see you made this change under different discussion item. Thanks!

arssher commented on broker/src/bin/bench.rs:
```diff
@@ -0,0 +1,175 @@
+use std::sync::atomic::{AtomicU64, Ordering};
```
Added as `-h` flag to the binary.

arssher commented:
> https://github.com/fullstorydev/grpcurl can be used as a client for the broker. For example, this query will subscribe for everything:

cool, added to `broker.md`

arssher commented:
- Added unit tests
- Added `broker.md`
- Went through reviews above
- Went ahead and replaced mpsc with `tokio::broadcast`, per suggestion above, and notify channels with rwlock per @petuhovskiy  suggestion above. Checked that performance is approximately the same, memory consumption is slightly higher on a lot of per-timeline subscriptions (but anyway decent) and obviously much less on many publishers. Saved about ~150 lines of code, and it got simpler.

Let's look at CI, but I consider this mostly ready for merge.

arssher commented:
Helm chart is being added at
https://github.com/neondatabase/helm-charts/pull/15

LizardWizzard commented on broker/src/bin/bench.rs:
```diff
@@ -0,0 +1,175 @@
+    let mut skipped: u64 = 0;
+    loop {
+        interval.tick().await;
+        // print!(
```
There are still other places with commented code.

arssher replied:
removed

LizardWizzard commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,515 @@
+        if let Some(subs) = shared_state.chans_to_timeline_subs.get(&ttid) {
+            // Err can't happen here, as rx is destroyed after removing from the
+            // map the last subscriber.
+            subs.chan.send(msg.clone()).unwrap();
```
Lets use `.expect(...)` here instead of unwrap

arssher replied:
done

LizardWizzard commented on broker/benches/neon_broker_bench.rs:
```diff
@@ -0,0 +1,174 @@
+use std::sync::atomic::{AtomicU64, Ordering};
```
NIT: the file can be named something like `timeline_pubsub` or something similar. `_bench` suffix is redundant to me since now fole lives under `/benches` and `neon_broker` too because everything happens under /broker

arssher replied:
Renamed to rps.rs, because that's what it really tests.

LizardWizzard commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,520 @@
+        );
+
+        // subscriber_all should receive both messages
+        assert!(matches!(subscriber_all.sub_rx.try_recv(), Ok(..)));
```
Lets check exact messages with `assert_eq`

arssher replied:
yeah, it also make the check much more concise

LizardWizzard commented on broker/src/bin/neon_broker.rs:
```diff
@@ -0,0 +1,520 @@
+        // subscriber_all should receive both messages
+        assert!(matches!(subscriber_all.sub_rx.try_recv(), Ok(..)));
+        assert!(matches!(subscriber_all.sub_rx.try_recv(), Ok(..)));
+        assert!(
+            matches!(subscriber_all.sub_rx.try_recv(), Err(err) if matches!(err, TryRecvError::Empty))
+        );
```
NIT: a bit more readable to me
```suggestion
        assert_eq!(
            subscriber_all.sub_rx.try_recv().unwrap_err(),
            TryRecvError::Empty
        );
```

vadim2404 commented on broker/benches/neon_broker_bench.rs:
```diff
@@ -0,0 +1,174 @@
+use neon_broker::proto::{SafekeeperTimelineInfo, SubscribeSafekeeperInfoRequest};
+use neon_broker::BrokerClientChannel;
+use neon_broker::DEFAULT_LISTEN_ADDR;
+use tokio::time::{self};
```
`use tokio::time;` — it seems that you had sth more when developed

LizardWizzard reviewed (commented):
Thanks for addressing the comments!

Looks mostly good to me. Minor comments remain, except the need for two extra allocations for each message

***

I would avoid neon prefix completely. In all the names of the files, variables etc. It makes things shorter and is consistent with naming in other components

arssher commented:
> I would avoid neon prefix completely.

I considered that, but unlike pageserver/safekeeper, `broker`  feels like too generic term to me. So from this stems binary name `neon_broker` and the rest.

petuhovskiy commented:
WDYT about logging every publisher and subscriber connection with `info!` level? I'd like to have logs like that:
```
INFO Started subscription for {sub_key}, ip={ip}
INFO Finished subscription for {sub_key}, ip={ip}
INFO Started publishing, ip={ip}
INFO Finished publishing, ip={ip}
```

arssher commented:
I'm +1, but I remember at some point we disabled logging loading timelines on safekeepers thinking that would it be too much log volume. Here it is roughly the same number :)

LizardWizzard commented:
> Perhaps storage_broker?

If we put control plane in this repo then all storage will be under /storage it may repeat `storage` but we have proxy here anyway. So maybe thats fine

arssher commented:
Renamed and added remote addr logging.

LizardWizzard commented on storage_broker/src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<hyper::HeaderMap>, Self::Error>> {
+        match self.get_mut() {
+            EitherBody::Left(b) => Pin::new(b).poll_trailers(cx).map_err(Into::into),
```
NIT: Using `.map(Into::into)` is rather confusing to me and I would say this is an anti pattern because its quite hard to read and find actual implementation that will be used. So typically `AnyError::from` is the preferred way.

Might be easier to get away with `Box::new(e) as _` this requires slightly changing bounds of A::Error/B::Error above or as I mentioned with `AnyError::from`

arssher replied:
Yes, all EitherBody is somewhat complicated. Tried suggested, but `AnyError::from` requires changing bounds to `A::Error:  std::error::Error + Send + Sync + 'static` in three places, a lot of boilerplate; the same for`Box::new` and `Box::new(e) as _` doesn't help.

LizardWizzard replied:
It is indeed a bit more verbose, but allows to peek into particular From impl without guessing around Into::into

LizardWizzard reviewed (commented):
Overall looks good to me. Minor comments. And still this problem with extra allocations

arssher commented:
> And still this problem with extra allocations

Well, this thingy gives several millions rps on my laptop. Performance/scalability here won't be a problem for a long time, and if it is, I suspect this extra allocation won't be the top priority.

LizardWizzard commented on storage_broker/benches/rps.rs:
```diff
@@ -0,0 +1,174 @@
+    /// Number of subscribers
+    #[clap(short = 's', long, value_parser, default_value_t = 1)]
+    num_subs: u64,
+    // Fake value to satisfy `cargo bench` passing it.
```
Gotcha, TIL

lassizci commented on storage_broker/src/metrics.rs:
```diff
@@ -0,0 +1,25 @@
+use once_cell::sync::Lazy;
+
+pub static NUM_PUBS: Lazy<IntGauge> = Lazy::new(|| {
+    register_int_gauge!("broker_active_publishers", "Number of publications")
```
Could we prefix these with `storage_broker_` instead of just plain ` broker`? We have `pageserver` metrics with `pageserver_` prefix, `safekeeper` metrics with `safekeeper_` prefix and `proxy` metrics with `proxy_` prefix. I won't insist, though, as it's not that big of a deal.

arssher replied:
Yes, renamed.
